File: website/src/lib/powerCurve.ts
export type PowerCurveConfig = {
  cutInSpeedMs: number;
  ratedSpeedMs: number;
  cutOutSpeedMs: number;
};

export type PowerCurveResult = {
  windSpeedMs: number | null;
  capacityFactor: number | null;
  operatingRegion:
    | "missing"
    | "below_cut_in"
    | "ramp_up"
    | "rated"
    | "above_cut_out";
};

export const DEFAULT_POWER_CURVE_CONFIG: PowerCurveConfig = {
  cutInSpeedMs: 3,
  ratedSpeedMs: 12,
  cutOutSpeedMs: 25,
};

export function estimateCapacityFactor(
  windSpeedMs: number | null,
  config: PowerCurveConfig = DEFAULT_POWER_CURVE_CONFIG
): PowerCurveResult {
  if (windSpeedMs === null || !Number.isFinite(windSpeedMs)) {
    return {
      windSpeedMs: null,
      capacityFactor: null,
      operatingRegion: "missing",
    };
  }

  const { cutInSpeedMs, ratedSpeedMs, cutOutSpeedMs } = config;

  if (windSpeedMs < cutInSpeedMs) {
    return {
      windSpeedMs,
      capacityFactor: 0,
      operatingRegion: "below_cut_in",
    };
  }

  if (windSpeedMs >= cutOutSpeedMs) {
    return {
      windSpeedMs,
      capacityFactor: 0,
      operatingRegion: "above_cut_out",
    };
  }

  if (windSpeedMs >= ratedSpeedMs) {
    return {
      windSpeedMs,
      capacityFactor: 1,
      operatingRegion: "rated",
    };
  }

  const numerator = windSpeedMs ** 3 - cutInSpeedMs ** 3;
  const denominator = ratedSpeedMs ** 3 - cutInSpeedMs ** 3;
  const capacityFactor = Math.max(0, Math.min(1, numerator / denominator));

  return {
    windSpeedMs,
    capacityFactor,
    operatingRegion: "ramp_up",
  };
}

export function formatCapacityFactor(value: number | null): string {
  if (value === null || !Number.isFinite(value)) {
    return "N/A";
  }

  return `${(value * 100).toFixed(1)}%`;
}

File: model_service/app/power_curve.py
from __future__ import annotations


def estimate_capacity_factor(
    wind_speed_ms: float | None,
    cut_in_ms: float = 3.0,
    rated_ms: float = 12.0,
    cut_out_ms: float = 25.0,
) -> float:
    if wind_speed_ms is None:
        return 0.0

    v = max(float(wind_speed_ms), 0.0)

    if v < cut_in_ms:
        return 0.0

    if cut_in_ms <= v < rated_ms:
        cf = ((v - cut_in_ms) / (rated_ms - cut_in_ms)) ** 3
        return round(min(max(cf, 0.0), 1.0), 6)

    if rated_ms <= v <= cut_out_ms:
        return 1.0

    return 0.0
